Starfish Tracing Model (#83)
getsentry/rfcs: text/0083-starfish-tracing-model.md

- Start Date: 2023-03-29
- RFC Type: feature
- RFC PR: [#83](https://github.com/getsentry/rfcs/pull/83)
- RFC Status: draft

# Summary

This RFC proposes a new tracing model for Sentry's performance product to better
address future product improvements.  The goal of this model is to allow
**storing entire traces without gaps**, support **dynamic sampling**, **indexing
of spans** and to **extract metrics from spans pre-sampling**.  This is an
evolution of the current transaction based approach.

# Motivation

Today Sentry has a strong concept of a "transaction" which appears in multiple parts of the
product.  It is both the transport of span data, the billable entity and the only indexed
part of the product experience.  This means that spans that exist outside of the transaction

> cannot be represented and it also means that spans within a transaction are not indexed itself.

Reviewer: ```suggestion
cannot be represented and it also means that spans within a transaction are not indexed.
```

The existing model has worked well for us to get started with evolving the Sentry errors

> product to capture performance traces, but it has restricted out ability to evolve the product

Reviewer: ```suggestion
product to capture performance traces, but it has restricted our ability to evolve the product
```

forward.  It has created some restrictions on the SDK technology side (from high

> memory pressure, payload size limits) and also has promoted a separate of transaction to
> span on the API layer which is untypical for tracing products.  It also has meant that Sentry

Reviewer: ```suggestion
memory pressure, payload size limits) and also has promoted a separation of transactions and
spans on the API layer which is untypical for tracing products.  It also has meant that Sentry
```

has challenges with accepting traces coming directly from an OpenTelemetry exporter as the
transaction concept is not a concept that OpenTelemetry has.

We want to set a future direction that enables more flexible product choices and that we can
move towards from our existing tracing model.  The goals are:

* Support capturing entire traces
* Have a data model story that allows us higher compatibility with Open Telemetry.  Specifically
  we want a model that would permit us to ingest Open Telemetry data right from an exporter
* Have a clear story for indexing and extracing metrics on a per-span level
* Unified spans and transactions from an SDK perspective
* Enable a path that allows clients to send 100% of spans outwards at least to a local aggregator,
  preferrably a remote relay
* To better and directly support dynamic sampling in the core tracing model

We want to lay out a better path forward that 

* capture entire traces
* browser tabs = a trace
* index and extract metrics on a span level
* clients send 100% of metrics
* dynamic sampling narrows in on traces

# Terms

> The new tracing model is an extension to our existing tracing model.  As such we try go adhere

Reviewer: ```suggestion
The new tracing model is an extension to our existing tracing model.  As such we try to adhere
```

    POST /api/analytics                              :190, 70ms
    GET /api/checkout/state                          :200, 500ms
    GET /api/checkout/cart                           :1100, 140ms
    <App/>                                           :1300, 180ms
    POST /api/analytics                              :done, 1450, 70ms
    GET /assistent/poll                              :done, 1450, 120ms
    POST /api/analytics                              :done, 1580, 70ms

    section API Service
    /api/checkout/state                              :crit, 240, 440ms
    cache.get session#58;[redacted]                  :360, 10ms
    db.query select from session                     :370, 20ms
    db.query select from user                        :390, 20ms
    db.query select from checkout                    :410, 20ms
    http.request GET http#58;//payments/poll  :450, 210ms
    thread.spawn refresh-checkout-cache              :done, 670, 220ms

    section Payment Service
    /poll                                            :crit, 470, 180ms
    db.query select from payment                     :490, 30ms
    db.query update payment                          :530, 60ms
```

> ## Span

Reviewer 2: I noticed a distinct lack of `span.status` (let's goooo).

Would it be prudent to introduce a basic status field though, similar to OpenTelemetry's? This will have 3 values:

- `ok` (successfully finished)
- `error` (error occurred during span)
- `unknown` (default value)

Author: FWIW I did not spec out the entire model here but yes, `status` should be on there.

> Spans are very similar to how the function today, but they get elevated to a more significant

Reviewer: ```suggestion
Spans behave very similar to how the function today, but they get elevated to a more significant
```

level.  They largely follow the general semantics in the wider tracing eco system.  To drive
our product ideas we are going to ensure that the quality of the spans is high and that they
provide at least the following pieces of information:

* `op`: defines the core operation that the span represents (eg: `db.query`)

> * `description`: the most significant description of what this operation is (eg: the database query).

Reviewer 2: I would heavily be in favor of re-naming this to `name`.

For those unfamiliar we currently have an inconsistency in the sentry schema where transactions have a `name` attribute and spans have a `description` (functionality identical).

I think most users are familiar with name as an attribute (closer to otel and other vendors as well then), and it would require less delta for us to approach.

We even alias span descriptions to names in the UI

<img width="908" alt="image" src="https://user-images.githubusercontent.com/18689448/228606691-b789a2ed-8d01-41e1-854d-58c09c14315b.png">

Author: I think if we want to get rid of `description` as the primary field then we need to rethink how we do the UX around some of those pieces. We currently stash into `description` what others for instance stash into `http.route` etc.

Reviewer 2: I don't think we need to get rid of description, simply just call it `name` instead of `description`. It can work the same exact way.

  The description also gets processed and cleaned in the process.

> * `trace_id`: a span relates to one trace by ID

Reviewer 2: Don't we also need a `span_id`? Otherwise how can we form a proper parent child relationship?

> * `parent_id`: a span optionally points back to a parent trace which could be from a different

Reviewer 2: ```suggestion
* `parent_span_id`: a span optionally points back to a parent trace which could be from a different
```

  service

> * `segment_id`: a span that is part of a segment, always refers back to it.

Reviewer 2: What is the type of a `segment_id`? Will it be 32 hex chars like a `trace_id`, 16 hex chars like a `span_id`, or something else entirely?

Author: It's the `span_id` of the segment it points to. Will clarify.

> * `is_segment`: when set to `true` this span is a segment.

Reviewer 2: If segment is meant to be either a transaction or an interaction, should we build it into the spec here?

so `segment: "transaction" | "interaction" | null`, where null says that the span is not a segment.

* `start_time` and `end_time` to give the span time information.

> * `tags`: a key/value pair of arbitrary tags that are set per-span.  Conceptionally however spans

Reviewer: ```suggestion
* `tags`: a key/value pair of arbitrary tags that are set per-span.  Conceptually however spans
```

> * `tags`: a key/value pair of arbitrary tags that are set per-span.  Conceptionally however spans

Reviewer 2: Currently there are a couple ways to store arbitrary data on the existing schema:

on transactions you have tags, contexts and extras (inherting from error events)

on spans you have tags and data. 

As per our spec, tags are restricted to string keys -> string values, and have strict size limits., while data is restricted to string keys, but can have arbitrary values (with more liberal size limits).

This has lead to most of the SDKs storing [arbitrary values in span data](https://develop.sentry.dev/sdk/performance/span-data-conventions/), and the OpenTelemetry conversion setting span attributes (otels arbitrary span data) in the data field as well.

Span data is now also used for semantic conventions, similar to how opentelemetry semantic conventions use otel span attributes. I would like to keep this behaviour. For example, the `span.data` with the key [`blocked_main_thread`](https://develop.sentry.dev/sdk/performance/span-data-conventions/#mobile) is used in the file io performance issue detector.

 I would like an arbitrary place to continue to store data to still exist, so that means we either move to saying this is `data`, or we update the limitations on `tags`.

Author: I agree that we should be keeping this. I won't have time yet to focus on the span schema so if you want, maybe you can sketch this out directly on the RFC with what the schema should look like?

  also inherit the tags of the segments they are contained in.
* `measurements`: are span specific metrics that are stored with the span.  There is a natural
  measurement of a span which is the `duration` which is automatically calculated from the difference
  of the end to the start timestamp.

Spans always belong to a trace, but not all spans belong to a segment. A span not belonging to a
segment are referred to as "detached" spans, spans that belong to a segment are "attached" spans.
A span must only be attached to a segment if it belongs to the same process and service.  Remote

> spans must never be attached to a segment.

Author: Again, that schema is not fully defined yet. However remote means from another process or service.

## Transaction

Previously transaction referred to the container that held spans.  In the future a "transaction"
refers to the name of a specific type of segment level span that describes a meaningful activity
that starts with a request, and results in some meaningful response.  By definition a span that
holds a transaction tag becomes a "segment".  Other than that, it's just a regular span.

## Segment

A segment is a special type of span that is the "logical" activity in a service.  For instance a
segment can be the endpoint implementation of an API request, it could be a task that is processed
by a task worker, it could be the navigation a user performs in a UI or a screen transition.
Conceptionally segments fall into two categories: "transactions" which are quite mechanical and
clearly defined operations and "interactions" which are user triggered operations.  The difference
is that an "interaction" has a human user as an actor in it, that might influence it, whereas a
"transaction" is unlikely to be interrupted once started.  A user for instance is quite likely to

> navigate again even before the previous interaction finished, whereas a task is more likely than no

Reviewer: ```suggestion
navigate again even before the previous interaction finished, whereas a task is more likely than not
```

> to conclude, even if what triggered the task is no longer interested in it's result.

Reviewer: ```suggestion
to conclude, even if what triggered the task is no longer interested in the result of the task.
```

product to capture performance traces, but it has restricted our ability to evolve the product
forward.  It has created some restrictions on the SDK technology side (from high
memory pressure, payload size limits) and also has promoted a separation of transactions and
spans on the API layer which is untypical for tracing products.  It also has meant that Sentry
has challenges with accepting traces coming directly from an OpenTelemetry exporter as the
transaction concept is not a concept that OpenTelemetry has.

We want to set a future direction that enables more flexible product choices and that we can
move towards from our existing tracing model.  The goals are:

* Support capturing entire traces
* Have a data model story that allows us higher compatibility with Open Telemetry.  Specifically
  we want a model that would permit us to ingest Open Telemetry data right from an exporter
* Have a clear story for indexing and extracing metrics on a per-span level
* Unified spans and transactions from an SDK perspective
* Enable a path that allows clients to send 100% of spans outwards at least to a local aggregator,

>   preferrably a remote relay

Reviewer 3: Which types of aggregation do you have in mind here ?
Is that the usual metrics extraction or something else?

* To better and directly support dynamic sampling in the core tracing model

We want to lay out a better path forward that 

* capture entire traces
* browser tabs = a trace
* index and extract metrics on a span level

> * clients send 100% of metrics

Reviewer 4: maybe this is in the TODO, but how does this proposal help with this "100% sending" or "memory-pressure/payload size" issue listed above? what about the current CPU overhead?

It is helpful for me to remember our own sampling limits, even after dynamic sampling has been implemented, are often set extremely low for performance reasons https://github.com/getsentry/sentry/blob/cc47128efd5a00e66b2475acc318ff57860fdf94/src/sentry/utils/sdk.py#L35

Reviewer 4: even with a local relay I don't understand how we could ever have metrics sampled enough to be useful for SLAs / Alerts where performance of the process is not severely degraded in medium throughput and above endpoints. It seems to me the only way to accomplish this is for metrics collection directly in the process.

Reviewer 4: ah maybe this is what 
>**extract metrics from spans pre-sampling**.

is going to address?

Reviewer 4: <img width="682" alt="Screenshot 2023-04-03 at 5 57 56 PM" src="https://user-images.githubusercontent.com/1976777/229659123-1518c48c-2a05-4e68-92d5-e92d9aecbded.png">
from a sentry staff engineer. I feel like for the type of customer that wants to use this product, _all_ of their endpoints are defined as "high perf backends".  have we talked to prospective customers about this?

    cache.get session#58;[redacted]                  :360, 10ms
    db.query select from session                     :370, 20ms
    db.query select from user                        :390, 20ms
    db.query select from checkout                    :410, 20ms
    http.request GET http#58;//payments/poll  :450, 210ms
    thread.spawn refresh-checkout-cache              :done, 670, 220ms

    section Payment Service
    /poll                                            :crit, 470, 180ms
    db.query select from payment                     :490, 30ms
    db.query update payment                          :530, 60ms
```

## Span

Spans behave very similar to how the function today, but they get elevated to a more significant
level.  They largely follow the general semantics in the wider tracing eco system.  To drive
our product ideas we are going to ensure that the quality of the spans is high and that they
provide at least the following pieces of information:

* `op`: defines the core operation that the span represents (eg: `db.query`)
* `description`: the most significant description of what this operation is (eg: the database query).

>   The description also gets processed and cleaned in the process.

Reviewer 3: How do we expect to use this? 
When we experimented with span search and when we built suspect spans one of the major issues was that all the useful information was stored unstructured in this field and all queries would have been regexes. It did not end well. 

@nikhars , if this is going to be a searchable field you may want to look into ngram indexes.
